File: skill/plugin/import-adapters/claude-adapter.ts
import { BaseImportAdapter } from './base-adapter.js';
import type {
  ImportSource,
  NormalizedFact,
  AdapterParseResult,
  ProgressCallback,
} from './types.js';
import fs from 'node:fs';
import os from 'node:os';

/**
 * Pattern for lines that start with a date prefix.
 * Claude memory entries sometimes have: [2026-03-15] - User prefers TypeScript
 */
const DATE_PREFIX_RE = /^\[(\d{4}-\d{2}-\d{2})\]\s*[-:]\s*/;

/**
 * Pattern for bullet-prefixed lines.
 */
const BULLET_PREFIX_RE = /^[-*\u2022]\s+/;

/**
 * Pattern for numbered list lines.
 */
const NUMBERED_PREFIX_RE = /^\d+[.)]\s+/;

/**
 * Patterns for classifying Claude memory entries by type.
 * Claude memories are already curated facts, so we use lighter heuristics.
 */
const TYPE_PATTERNS: Array<{ pattern: RegExp; type: NormalizedFact['type'] }> = [
  { pattern: /\bprefers?\b/i, type: 'preference' },
  { pattern: /\blikes?\b/i, type: 'preference' },
  { pattern: /\bdislikes?\b/i, type: 'preference' },
  { pattern: /\bfavorite\b/i, type: 'preference' },
  { pattern: /\bfavourite\b/i, type: 'preference' },
  { pattern: /\bavoids?\b/i, type: 'preference' },
  { pattern: /\bdecided\b/i, type: 'decision' },
  { pattern: /\bchose\b/i, type: 'decision' },
  { pattern: /\bwants? to\b/i, type: 'goal' },
  { pattern: /\bplans? to\b/i, type: 'goal' },
  { pattern: /\bgoal\b/i, type: 'goal' },
  { pattern: /\baims? to\b/i, type: 'goal' },
  { pattern: /\bworking on\b/i, type: 'context' },
  { pattern: /\bcurrently\b/i, type: 'context' },
  { pattern: /\bproject\b/i, type: 'context' },
];

function classifyMemory(text: string): NormalizedFact['type'] {
  for (const { pattern, type } of TYPE_PATTERNS) {
    if (pattern.test(text)) return type;
  }
  return 'fact';
}

export class ClaudeAdapter extends BaseImportAdapter {
  readonly source: ImportSource = 'claude';
  readonly displayName = 'Claude';

  async parse(
    input: { content?: string; file_path?: string },
    onProgress?: ProgressCallback,
  ): Promise<AdapterParseResult> {
    const warnings: string[] = [];
    const errors: string[] = [];

    let content: string;

    if (input.content) {
      content = input.content;
    } else if (input.file_path) {
      try {
        const resolvedPath = input.file_path.replace(/^~/, os.homedir());
        content = fs.readFileSync(resolvedPath, 'utf-8');
      } catch (e) {
        errors.push(`Failed to read file: ${e instanceof Error ? e.message : 'Unknown error'}`);
        return { facts: [], warnings, errors };
      }
    } else {
      errors.push(
        'Claude import requires either content (pasted text) or file_path. ' +
        'Copy your memories from Claude: Settings -> Memory -> select all and copy.',
      );
      return { facts: [], warnings, errors };
    }

    // Claude memory export is plain text, one fact per line.
    // Sometimes with date prefixes like [2026-03-15] - User prefers TypeScript.
    // Sometimes with bullet points or numbered lists.
    return this.parseMemoriesText(content.trim(), warnings, errors, onProgress);
  }

  /**
   * Parse Claude memories — plain text, one memory per line.
   */
  private parseMemoriesText(
    content: string,
    warnings: string[],
    errors: string[],
    onProgress?: ProgressCallback,
  ): AdapterParseResult {
    // Split by newlines and filter
    const lines = content.split('\n')
      .map((line) => line.trim())
      .filter((line) => line.length > 0)
      // Skip common header lines
      .filter((line) => !/^(?:memories?|claude memories?|my memories?|saved memories?):?\s*$/i.test(line));

    if (onProgress) {
      onProgress({
        current: 0,
        total: lines.length,
        phase: 'parsing',
        message: `Parsing ${lines.length} Claude memories...`,
      });
    }

    const rawFacts: Partial<NormalizedFact>[] = lines.map((line, i) => {
      let cleaned = line;
      let timestamp: string | undefined;

      // Extract date prefix if present
      const dateMatch = cleaned.match(DATE_PREFIX_RE);
      if (dateMatch) {
        timestamp = dateMatch[1];
        cleaned = cleaned.replace(DATE_PREFIX_RE, '');
      }

      // Strip bullet/numbering markers
      cleaned = cleaned
        .replace(BULLET_PREFIX_RE, '')
        .replace(NUMBERED_PREFIX_RE, '')
        .trim();

      const type = classifyMemory(cleaned);

      return {
        text: cleaned.slice(0, 512),
        type,
        // Claude memories are already curated -- default to importance 6
        importance: 6,
        source: 'claude' as ImportSource,
        sourceId: `claude-memory-${i}`,
        sourceTimestamp: timestamp,
        tags: ['claude-memory'],
      };
    });

    const { facts, invalidCount } = this.validateFacts(rawFacts);

    if (invalidCount > 0) {
      warnings.push(`${invalidCount} memories had invalid/empty text and were skipped`);
    }

    return {
      facts,
      warnings,
      errors,
      source_metadata: {
        format: 'memories-text',
        total_lines: lines.length,
      },
    };
  }
}

File: python/src/totalreclaw/import_adapters/claude_adapter.py
"""
Claude import adapter -- parses plain text memories (one per line, optional date prefix).

Ported from skill/plugin/import-adapters/claude-adapter.ts
"""

import os
import re
from typing import Optional, Callable, List

from .base_adapter import BaseImportAdapter
from .types import AdapterParseResult, ConversationChunk


# Pattern for lines that start with a date prefix.
# Claude memory entries sometimes have: [2026-03-15] - User prefers TypeScript
DATE_PREFIX_RE = re.compile(r'^\[(\d{4}-\d{2}-\d{2})\]\s*[-:]\s*')

# Pattern for bullet-prefixed lines.
BULLET_PREFIX_RE = re.compile(r'^[-*\u2022]\s+')

# Pattern for numbered list lines.
NUMBERED_PREFIX_RE = re.compile(r'^\d+[.)]\s+')

# Maximum messages per conversation chunk for LLM extraction.
CHUNK_SIZE = 20


class ClaudeAdapter(BaseImportAdapter):
    source = 'claude'
    display_name = 'Claude'

    def parse(
        self,
        *,
        content: Optional[str] = None,
        file_path: Optional[str] = None,
        on_progress: Optional[Callable] = None,
    ) -> AdapterParseResult:
        warnings: List[str] = []
        errors: List[str] = []

        if content:
            raw = content
        elif file_path:
            try:
                resolved = os.path.expanduser(file_path)
                with open(resolved, 'r', encoding='utf-8') as f:
                    raw = f.read()
            except Exception as e:
                errors.append(f'Failed to read file: {e}')
                return AdapterParseResult(
                    facts=[], chunks=[], total_messages=0,
                    warnings=warnings, errors=errors,
                )
        else:
            errors.append(
                'Claude import requires either content (pasted text) or file_path. '
                'Copy your memories from Claude: Settings -> Memory -> select all and copy.',
            )
            return AdapterParseResult(
                facts=[], chunks=[], total_messages=0,
                warnings=warnings, errors=errors,
            )

        # Claude memory export is plain text, one fact per line.
        return self._parse_memories_text(raw.strip(), warnings, errors, on_progress)

    def _parse_memories_text(
        self,
        content: str,
        warnings: List[str],
        errors: List[str],
        on_progress: Optional[Callable] = None,
    ) -> AdapterParseResult:
        """
        Parse Claude memories -- plain text, one memory per line.
        Returns conversation chunks for LLM extraction (no pattern matching).

        Each line is cleaned (date prefixes, bullets, numbers stripped) and
        grouped into chunks for the LLM to process.
        """
        # Split by newlines and filter
        lines = [line.strip() for line in content.split('\n') if line.strip()]
        # Skip common header lines
        header_re = re.compile(
            r'^(?:memories?|claude memories?|my memories?|saved memories?):?\s*$',
            re.IGNORECASE,
        )
        lines = [line for line in lines if not header_re.match(line)]

        if on_progress:
            on_progress({
                'current': 0,
                'total': len(lines),
                'phase': 'parsing',
                'message': f'Parsing {len(lines)} Claude memories...',
            })

        # Clean each line: extract date, strip formatting
        cleaned_entries: List[dict] = []
        for line in lines:
            cleaned = line
            timestamp = None

            # Extract date prefix if present
            date_match = DATE_PREFIX_RE.match(cleaned)
            if date_match:
                timestamp = date_match.group(1)
                cleaned = DATE_PREFIX_RE.sub('', cleaned)

            # Strip bullet/numbering markers
            cleaned = BULLET_PREFIX_RE.sub('', cleaned)
            cleaned = NUMBERED_PREFIX_RE.sub('', cleaned)
            cleaned = cleaned.strip()

            if len(cleaned) >= 3:
                cleaned_entries.append({'text': cleaned, 'timestamp': timestamp})

        # Group memories into chunks of CHUNK_SIZE for efficient LLM extraction
        chunks: List[ConversationChunk] = []
        for i in range(0, len(cleaned_entries), CHUNK_SIZE):
            batch = cleaned_entries[i:i + CHUNK_SIZE]

            # Use the timestamp from the first entry in the batch (if available)
            batch_timestamp = None
            for entry in batch:
                if entry['timestamp']:
                    batch_timestamp = entry['timestamp']
                    break

            chunks.append(ConversationChunk(
                title=f'Claude memories ({i + 1}-{min(i + CHUNK_SIZE, len(cleaned_entries))})',
                messages=[{'role': 'user', 'text': entry['text']} for entry in batch],
                timestamp=batch_timestamp,
            ))

        return AdapterParseResult(
            facts=[],
            chunks=chunks,
            total_messages=len(cleaned_entries),
            warnings=warnings,
            errors=errors,
            source_metadata={
                'format': 'memories-text',
                'total_lines': len(lines),
                'chunks_count': len(chunks),
            },
        )
